### ISSUE TITLE
When input is empty but not None, `torch.autograd.grad` populates .grad of leaf Tensors

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#47061]
## 🐛 Bug

When input is empty but not None, `torch.autograd.grad` populates .grad of leaf Tensors 

## To Reproduce

The code below fails:
```
x = torch.tensor([1.0], requires_grad=True);
grad = torch.autograd.grad(2*x, [], grad_outputs=torch.tensor([1.0]))
assert(x.grad is None)
```

## Expected behavior

Gradients should not accumulate into leafs for `torch.autograd.grad`

## Environment

PyTorch version: 1.8.0a0+2518abb
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 8 (Core) (x86_64)
GCC version: (GCC) 8.3.1 20191121 (Red Hat 8.3.1-5)
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.1
[pip3] torch==1.8.0a0
[conda] blas                      1.0                         mkl
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.1.0            py38h23d657b_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.1           py38hbc911f0_0
[conda] numpy-base                1.19.1           py38hfa32c7d_0
[conda] torch                     1.8.0a0                   dev_0    <develop>

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

**Comment by albanD** [comment]
This behavior is definitely wrong.
We should definitely change this to be an error, or just return an empty tuple without calling into the autograd.

@gchanan do you think we need to go through a full deprecation with warning then error? Or we can make it an error right now?
